core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java

```diff
@@ -103,27 +97,37 @@ private void finishResolve(ClientRequestContext ctx, Endpoint endpoint,
 
     private void executeWithIpAddr(ClientRequestContext ctx, Endpoint endpoint, String ipAddr,
                                    HttpRequest req, DecodedHttpResponse res) {
-        final String host = extractHost(ctx, req, endpoint);
-        final PoolKey poolKey = new PoolKey(host, ipAddr, endpoint.port(), ctx.sessionProtocol());
-        final Future<Channel> channelFuture = factory.pool(ctx.eventLoop()).acquire(poolKey);
-
-        if (channelFuture.isDone()) {
-            finishExecute(ctx, poolKey, channelFuture, req, res);
+        final EventLoop eventLoop = ctx.eventLoop();
+        if (eventLoop.inEventLoop()) {
+            executeWithIpAddr0(ctx, endpoint, ipAddr, req, res);
         } else {
-            channelFuture.addListener(
-                    (Future<Channel> future) -> finishExecute(ctx, poolKey, future, req, res));
+            eventLoop.execute(() -> executeWithIpAddr0(ctx, endpoint, ipAddr, req, res));
         }
     }
 
-    private void finishExecute(ClientRequestContext ctx, PoolKey poolKey, Future<Channel> channelFuture,
-                               HttpRequest req, DecodedHttpResponse res) {
-        if (channelFuture.isSuccess()) {
-            final Channel ch = channelFuture.getNow();
-            invoke0(ch, ctx, req, res, poolKey);
+    private void executeWithIpAddr0(ClientRequestContext ctx, Endpoint endpoint, String ipAddr,
+                                    HttpRequest req, DecodedHttpResponse res) {
+        assert ctx.eventLoop().inEventLoop();
+
+        final String host = extractHost(ctx, req, endpoint);
+        final int port = endpoint.port();
+        final SessionProtocol protocol = ctx.sessionProtocol();
+        final HttpChannelPool pool = factory.pool(ctx.eventLoop());
+
+        final PoolKey key = new PoolKey(host, ipAddr, port);
+        final PooledChannel pooledChannel = pool.acquireNow(protocol, key);
+        if (pooledChannel != null) {
+            invoke0(pooledChannel,ctx, req, res);
         } else {
-            final Throwable cause = channelFuture.cause();
-            handleEarlyRequestException(ctx, req, cause);
-            res.close(cause);
+            pool.acquire(protocol, key).whenComplete((newPooledChannel, cause) -> {
```

Comment:
Just curious. Don't we need to use `handle` instead of `whenComplete` in order to avoid creating `CompletionException` internally?

Reply:
It's OK because we did not chain after calling `whenComplete`, but let me switch to `handle` anyway just in case we add more in the future.